libraries/AC_Avoidance/AP_OADatabase.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Common/Location.h>
+
+#if !HAL_MINIMIZE_FEATURES
+#include <AP_Param/AP_Param.h>
+#include <AP_HAL/utility/RingBuffer.h>
+
+class AP_OADatabase {
+public:
+
+    AP_OADatabase();
+
+    /* Do not allow copies */
+    AP_OADatabase(const AP_OADatabase &other) = delete;
+    AP_OADatabase &operator=(const AP_OADatabase&) = delete;
+
+    // get singleton instance
+    static AP_OADatabase *get_singleton() {
+        return _singleton;
+    }
+
+    struct OA_DbItem {
+        // TODO: turn this into Vector2Int to save memory
+        Location loc;
+        uint32_t timestamp;
+    };
+
+    void init();
+    void update();
+
+    // push a location into the database
+    void queue_push(const Location &loc, const uint32_t timestamp);
+
+    bool healthy() const { return _initialised; }
+
+    // fetch an item in database. Undefined result when i >= _database.count.
+    const OA_DbItem get_item(uint32_t i) const { return _database.items[i]; }
+
+
+    float get_accuracy() const { return _database.filter_m; }
+    uint32_t database_count() const { return _database.count; }
+    uint32_t queue_count() const { return _queue.items->available(); }
+
+    static const struct AP_Param::GroupInfo var_info[];
+
+private:
+    // empty queue and try and put into database
+    void process_queue();
+
+    // initialise
+    void init_queue();
+    void init_database();
+
+    // check queue and database sizes and adjust filter criteria to optimize use
+    void optimize_db_filter();
+
+    // database item management
+    void database_item_add(const OA_DbItem &item);
+    void database_item_refresh(const uint32_t index, const uint32_t timestamp);
+    void database_item_remove(const uint32_t index);
+    void database_items_remove_all_expired();
+
+    bool is_close_to_item_in_database(const uint32_t index, const OA_DbItem &item) const;
+
+    AP_Int32        _queue_size_param;                      // queue size
+    AP_Int32        _database_size_param;                   // db size
+    AP_Int8         _database_expiry_seconds;               // objects expire after this timeout
+    bool _initialised;
+
+    struct {
+        ObjectBuffer<OA_DbItem> *items;                     // thread safe incoming queue of points from proximity sensor to be put into database
+        uint32_t        size;                               // cached value of _queue_size_param.
+        HAL_Semaphore   sem;                                // semaphore for multi-thread use of queue
+    } _queue;
+
+    struct {
+        OA_DbItem       *items;                             // THE DATABASE
+        uint32_t        optimize_last_ms;
+        float           filter_m                = 1.0f;     // object avoidance database optimization level radius. Min distance between each fence point. Larger means lower resolution
```

Comment:
It's not a blocker but I'm not sure that "filter_m" is a great name.  I think "radius" should be included in the name somehow.  Normally we don't add the units at the end if it's SI already.  Anyway.. not a huge deal but perhaps at some point we can come up with a better name.

Reply:
active_radius or current_radius is probably better. next PR for that stuff